Add TrieMap

julienrf opened the pull request:
This is a straightforward port of the existing `TrieMap` collection.

Ichoran commented on src/main/scala/strawman/collection/concurrent/TrieMap.scala:
```diff
@@ -0,0 +1,1146 @@
+}
+
+
+private[collection] final class LNode[K, V](final val listmap: immutable.ListMap[K, V])
```
This whole thing is broken because it doesn't respect custom equals.  We need to replace it with a simple `List` that has the appropriate operations but defers to the appropriate `equals`.

Ichoran commented on src/main/scala/strawman/collection/concurrent/TrieMap.scala:
```diff
@@ -0,0 +1,1146 @@
+      new TNode(k, v, ct.computeHash(k)) // create it tombed so that it gets compressed on subsequent accesses
+    }
+  }
+  def get(k: K) = listmap.get(k)
```
This line in particular fails to honor a custom `equal`.

Ichoran commented on src/main/scala/strawman/collection/concurrent/TrieMap.scala:
```diff
@@ -0,0 +1,1146 @@
+  def this(k1: K, v1: V, k2: K, v2: V) = this(immutable.ListMap(k1 -> v1, k2 -> v2))
+  def inserted(k: K, v: V) = new LNode(listmap + ((k, v)))
+  def removed(k: K, ct: TrieMap[K, V]): MainNode[K, V] = {
+    val updmap = listmap - k
```
This line uses the default equality and thus is wrong.

Ichoran commented on src/main/scala/strawman/collection/mutable/Map.scala:
```diff
@@ -65,6 +65,18 @@ trait MapOps[K, V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
+    */
+  def remove(key: K): Option[V] = {
+    val r = get(key)
+    this -= key
```
This is not a very efficient way to do this.  You should at least check whether `r` is empty before subtracting `key`, and better is to do it in a single operation.  (That's kind of the point of this method--it's supposed to be the efficient low-level operation in Java maps.)

julienrf replied:
Do you have an idea of how to implement it generically? I think we want to override this default implementation in each specific Map implementation.

julienrf commented:
Indeed I copy-pasted the current implementation and modified very few things to make it fit into the strawman.

I added a separate commit to fix scala/bug#10481.

julienrf commented:
I rebased to master and enabled the existing junit tests for `TrieMap`
